bookkeeper-server/src/main/java/org/apache/bookkeeper/proto/PacketProcessorBase.java

```diff
@@ -175,7 +175,9 @@ public void run() {
             if (request instanceof BookieProtocol.ReadRequest) {
                 requestProcessor.onReadRequestFinish();
             }
-            if (request instanceof BookieProtocol.AddRequest) {
+            if (request instanceof BookieProtocol.ParsedAddRequest) {
+                ((BookieProtocol.ParsedAddRequest) request).release();
+                request.recycle();
                 requestProcessor.onAddRequestFinish();
```

Comment:
Can the release of ParsedAddRequest be executed in the recycle method? It seems that these two methods are always used together, and there is a forced switch here, which seems a bit awkward.

You can look at org.apache.bookkeeper.proto.BookieProtocol.AddRequest#recycle method
```java
        @Override
        public void recycle() {
            ledgerId = -1;
            entryId = -1;
            masterKey = null;
            ReferenceCountUtil.safeRelease(data);
            data = null;
            recyclerHandle.recycle(this);
        }
```

Reply:
It can't, we shouldn't release the data in the recycle.
In the normal case, it will recycle the request, but the data is already free.

See: https://github.com/apache/bookkeeper/blob/7b5b6b240ca7551da266903078f2dd2ba2906b96/bookkeeper-server/src/main/java/org/apache/bookkeeper/proto/WriteEntryProcessor.java#L131

Reply:
I saw it, sorry I misread it earlier. Nice work. Thanks.